src/aids-env.adb

```diff
@@ -1,58 +1,85 @@
 with Ada.Text_IO; use Ada.Text_IO;
+with Ada.Strings.Fixed;
 
 package body Aids.Env is
+    Function Find(Env   : in Typ; 
+                  Key   : in String;
+                  Found : in out Boolean
+                 ) return String is
+    Begin
+        Found := Env.Contains( Key );
+        Return (if not Found then "" else Env(Key));
+    End Find;
+
     function Index_Of(Line: in String; X: in Character; Result: out Integer) return Boolean is
+        Index : Natural renames Ada.strings.Fixed.Index(
+           Source  => Line,      -- The String we're searching,
+           Pattern => (1 => X),  -- the *Substring* to search for;
+           From    => Line'First -- the point we start from.
+          );
     begin
-        for Index in Line'Range loop
-            if Line(Index) = X then
-                Result := Index;
-                return True;
-            end if;
-        end loop;
-        return False;
+        Return Found : constant Boolean := Index in Positive do
+            Result:= Index;
+        End return;
     end;
 
     function Slurp(File_Path: String) return Env_Hashed_Map.Map is
-        Result: Env_Hashed_Map.Map;
         File: File_Type;
         Line_Number: Integer := 1;
     begin
-        Open(File => File,
-             Mode => In_File,
-             Name => File_Path);
-        while not End_Of_File(File) loop
-            declare
-                Line: String := Get_Line(File);
-                Index : Integer;
-            begin
-                if Index_Of(Line, '=', Index) then
-                    declare
-                        Key : Unbounded_String := To_Unbounded_String(Line(Line'First..Index-1));
-                        Value : Unbounded_String := To_Unbounded_String(Line(Index+1..Line'Last));
-                    begin
-                        Env_Hashed_Map.Insert(Result, Key, Value);
-                    end;
-                else
-                    raise Syntax_Error with (File_Path 
-                                             & ":" 
-                                             & Integer'Image(Line_Number)
-                                             & ": Expected separator `=`");
-                end if;
+        Return Result: Env_Hashed_Map.Map do
```

Comment:
I had no idea return statements could be done like this. Very cool.

Reply:
They are, the two great things about the extended return are (1) that it gives you a "context" to your result wherein to operate, and (2) it can simplify your logic.

The `Index_Of` rewrite is a good, tiny example; where we've already determined the return-value, but need to satisfy the output-parameter:

     function Index_Of(Line: in String; X: in Character; Result: out Integer) return Boolean is
        Index : Natural renames Ada.strings.Fixed.Index(
           Source  => Line,      -- The String we're searching,
           Pattern => (1 => X),  -- the *Substring* to search for;
           From    => Line'First -- the point we start from.
          );
     begin
        Return Found : constant Boolean := Index in Positive do
            Result:= Index;
        End return;
     end;

Another good use is something like dealing with fix-ups, initializations, and sometimes just OOP.

    -- Default commands
    -- Command_Map:  String -> Command.
    -- Command: Not null Access Procedure( Object : in out Virtual_Machine );
    Function Init return Command_Map is
    Begin
       Return Result : Command_Map do
         Result.Include( "Start", Start_VM'Access );
         Result.Include( "Stop", Stop_VM'Access );
         --…
       end return;
    End Init;

Reply:
Extended return statements were added so you can return objects of limited types in-place, and with a bit of initialization code